Benchmark qps

liyaoxuan opened the pull request:
-  implement throughput control in redis-benchmark using producer/consumer model  
    * use `--qps` arg to control throughput
- change the latency computation logic to make it closer to the real scene (from user's perspective)

caipengbo commented on src/redis-benchmark.c:
```diff
@@ -628,16 +635,11 @@ static void writeHandler(aeEventLoop *el, int fd, void *privdata, int mask) {
-            }
+            long long timestamp = popItem();
+            if (timestamp == -1) return;
+            else  c->start = timestamp;
```
I think there's a problem here. This timestamp is not the actual timestamp sent to the Server.

caipengbo commented on src/redis-benchmark.c:
```diff
@@ -980,6 +982,40 @@ static void initBenchmarkThreads() {
+    pthread_mutex_init(&itemQueue.mutex, NULL);
+    itemQueue.size = size;
+    if (itemQueue.queue) zfree(itemQueue.queue);
+    itemQueue.queue = zmalloc(size * sizeof(long long));
```
If the `size`(config.requests) is large, it will allocate a lot of memory.

caipengbo commented on src/redis-benchmark.c:
```diff
@@ -605,17 +622,24 @@ static void writeHandler(aeEventLoop *el, int fd, void *privdata, int mask) {
+        if (config.qps) {
+            long long timestamp = popItem();
+            if (timestamp == -1) return;
+            else  c->start = timestamp;
```
I think there's a problem here. This timestamp is not the actual timestamp sent to the Server.

liyaoxuan replied:
yes, I have modified the code to use the actual timestamp.

caipengbo commented on src/redis-benchmark.c:
```diff
@@ -958,6 +982,48 @@ static void initBenchmarkThreads() {
+    pthread_mutex_init(&itemQueue.mutex, NULL);
+    itemQueue.size = size;
+    if (itemQueue.queue) zfree(itemQueue.queue);
+    itemQueue.queue = zmalloc(size * sizeof(long long));
```
If the size is large, it will allocate a lot of memory

caipengbo commented on src/redis-benchmark.c:
```diff
@@ -1408,6 +1502,13 @@ int parseOptions(int argc, char **argv) {
         if (!strcmp(argv[i],"-c")) {
             if (lastarg) goto invalid;
             config.numclients = atoi(argv[++i]);
+        } else if (!strcmp(argv[i], "--dist")) {
```
I don't think it's necessary to add this argument, the granularity is too fine. We just need to make sure the commands are sent smoothly.

liyaoxuan replied:
it's true this argument is unnecessary at present.

caipengbo commented on src/redis-benchmark.c:
```diff
@@ -1408,6 +1502,13 @@ int parseOptions(int argc, char **argv) {
+        } else if (!strcmp(argv[i], "--dist")) {
+            if (lastarg) goto invalid;
+            config.distribution = atoi(argv[++i]);
+        } else if (!strcmp(argv[i], "--qps")) {
```
I think it is more appropriate to use `--throughput` because `redis-benchmark` uses this word

liyaoxuan replied:
I think there is a little bit difference between qps and throughput. Qps here refers to the requests load created by benchmark (what we want to control), throughput refers to the actual processed requests per second. So I think qps is more appropriate here.

caipengbo commented on src/redis-benchmark.c:
```diff
@@ -1041,6 +1114,27 @@ static void *execBenchmarkThread(void *ptr) {
+            for (int i = 0; i < config.numclients; ++i) {
+                pushItem(timenow + i * USECOND / qps);
+            }
+            config.requests_emitted += config.numclients;
```
The `pipeline` case is not considered.

caipengbo commented on src/redis-benchmark.c:
```diff
@@ -1041,6 +1114,27 @@ static void *execBenchmarkThread(void *ptr) {
+            for (int i = 0; i < config.numclients; ++i) {
+                pushItem(timenow + i * USECOND / qps);
+            }
+            config.requests_emitted += config.numclients;
```
The pipeline case is not considered.

liyaoxuan replied:
yes, I will modify it in the next pr

yoav-steinberg commented on src/redis-benchmark.c:
```diff
@@ -125,8 +126,20 @@ static struct config {
     int enable_tracking;
     pthread_mutex_t liveclients_mutex;
     pthread_mutex_t is_updating_slots_mutex;
+    int requests_emitted;
```
I think this needs to be re-zeroed per call to `benchmark()`, In any case running the code like this:
`src/redis-benchmark --qps 2000 -n 10000`
Never exists because the second "mbulk_ping" benchmark never finishes.

liyaoxuan replied:
Yes, it should be re-zeroed

yoav-steinberg commented on src/redis-benchmark.c:
```diff
@@ -605,17 +622,24 @@ static void writeHandler(aeEventLoop *el, int fd, void *privdata, int mask) {
-        c->start = ustime();
+        if (config.qps) {
+            long long timestamp = popItem();
+            if (timestamp == -1) return;
```
There might be a busy loop here, returning without performing any write will immediately call the write handler again.

liyaoxuan replied:
I thought it would be processed in the next loop... So from my understanding, it may cause the calculated latency of other requests larger than actual value?
I tried to fix it by deleting and recreating the writing event, but it seems that the latency increased instead...

yoav-steinberg replied:
> I thought it would be processed in the next loop... So from my understanding, it may cause the calculated latency of other requests larger than actual value? 

It will be processed in the next loop, but as long as `popItem` returns `-1` you'll return and immediately call the write handler again. This will make the client thread work at 100% cpu even though there's nothing to do. I think `popItem` should block until there's more data to be sent. Does this make sense?

> I tried to fix it by deleting and recreating the writing event, but it seems that the latency increased instead...

I'm not sure you can know when to recreate the write event.

liyaoxuan replied:
I blocked  `popitem` when there is no item to fetch. But after that the p50, p95 and p99 latency increased from below 1ms to 20+ms. After I read the code of ae library, I guess the reason is when `popItem` is blocked, `aeProcessEvents` in **ae.c** is also blocked and thus not able to process the coming received requests timely. If that holds, the p99 latency will increase as QPS drops because the interval between two "producing activity" increased, and the experiments result confirmed my hypothesis. So I think it is not suitable to block `popItem` if I understand the writeHandler and ae event process correctly.
I'm now trying another method to use `aeTimeEvent` to control the emit rate which will solve the problem of busy waiting and blocking if works well.

yoav-steinberg replied:
>  I guess the reason is when `popItem` is blocked, `aeProcessEvents` in **ae.c** is also blocked and thus not able to process the coming received requests timely. ... I'm now trying another method to use `aeTimeEvent` to control the emit rate which will solve the problem of busy waiting and blocking if works well.

@liexusong This makes sense, if you have more than one client per thread then waiting on the queue might delay the other client's read/write handlers.

> I'm now trying another method to use aeTimeEvent to control the emit rate which will solve the problem of busy waiting and blocking if works well.

It'll be interesting to see if this works or has it own performance implications. I think that if we'll add a timer event per message it'll be problematic, if there's a good way to periodically add a timer event once we detect we need to slow down the QPS then this might work nicely. Alternatively you might want to try replacing your queue with an `anetPipe`, this way you can integrate it into the event loop without causing the delays you noticed.

yoav-steinberg reviewed (commented):
This seems like a good idea. Nice work. Added some questions.

itamarhaber commented:
I'm all for this one, but I'd like to have @filipecosta90's 2c as well.

liyaoxuan commented:
yes，i'm working to solve the last problem about writeHandler

liyaoxuan commented:
@ranshid yes, it's exactly the two methods I'm working on! I have implemented the first method which is to control every request using time event. The result shows that it can achieve fine-grained control when qps is low but not works well when qps grows. The reason is that `aeTimeEvent` is controlled in millisecond which is not able to handle when qps is large enough that the time interval between two request is less than 1ms. So I tried to set different control granularity for low qps and high qps (the low and high here is relavant to number of clients, i.e. qps per client), when the qps per client is small, the control granularity is 1 request, else we control the qps every 10 requests, this partially solves the problem, but is unstable when qps is very high. I have updated the code and will try the second method : )

liyaoxuan commented:
> @liyaoxuan I can see that you changed the producer/consumer design and went with a "per-client" qps control approach. This if fine, but I would like to ask again regarding the burstiness effect when benchmark is using many clients. for example assume the benchmark is configured to use 20K clients, since we only limiting the qps per client there would still be bursts of 20K requests issued. maybe it would be better to rate limit the qps per thread?

Yes, Burstiness effect is a critical problem of "per-client" control. I've implemented "per thread" control and the experiment result shows that burstiness effect is reduced. : )

yoav-steinberg reviewed (commented):
Nice work.
One minor note: please use block comments (`/* */`) instead of line comments (`//`) as we don't have line comments in the code base.

My other big question is that I'm a bit worried about the fixed `resume_interval` and `control_granularity`. Can't we track the qps per thread in runtime and then pause the relevant clients as they pass the desired qps. I haven't thought about this too much, but I'd imagine it's possible to find some algorithm where we calculate the resume_interval and granularity on the go and update them based on actual client qps. A minor note is that this will probably require float calculations which is fine.

liyaoxuan commented:
> Can't we track the qps per thread in runtime and then pause the relevant clients as they pass the desired qps?

If I understand it correctly, setting granularity to "qps per thread" and interval to 1s will achieve such purpose.

In general, the role of `control_granularity` here is to control the distribution of requests sent. Smaller `control_granularity` will make the requests smoother. However as the `aeTimeEvent` measures time in `ms`, there exists a lower boundary of `control_granularity` that the interval between finishing a group of requests and next resume is larger than 1ms when control is needed.

> but I'd imagine it's possible to find some algorithm where we calculate the resume_interval and granularity on the go and update them based on actual client qps.

I agree that it should be better to adjust the two parameters dynamically. In my opinion, the main idea of the algorithm is that decreasing granularity and interval when the pause time is larger than a threshold (e.g. 2ms?) and increasing when control is not triggered. I'm not sure about the effect and will try it later.

liyaoxuan commented:
I've finished the dynamic version as I described, but the experiment result is not as good as expected, actually the fixed version works better.
the experiments result is shown below
![image](https://user-images.githubusercontent.com/88424383/140933554-f634b193-84ba-4605-9804-dde6f4b37ece.png)

So it seems that fixed version is a better choice? Or I implemented the dynamic version in a wrong way...

ranshid commented on src/redis-benchmark.c:
```diff
@@ -971,6 +1040,40 @@ static void startBenchmarkThreads() {
+    int qps_per_thread = config.qps / num_threads;
+    config.paused_clients = zmalloc(num_threads * sizeof(list *));
+    config.last_resume_time = zmalloc(num_threads * sizeof(long long));
+    config.count = zmalloc(num_threads * sizeof(int));
```
We should make sure the config.count and config .pause are initialized
either use zcalloc or in the following for loop

liyaoxuan replied:
fixed

ranshid commented on src/redis-benchmark.c:
```diff
@@ -971,6 +1040,40 @@ static void startBenchmarkThreads() {
+        qps_per_thread = 1;
+    }
+    for (int i = 0; i < num_threads; ++i) {
+        config.paused_clients[i] = zmalloc(sizeof(list));
```
```suggestion
        config.paused_clients[i] = listCreate();
```

yoav-steinberg commented:
> So it seems that fixed version is a better choice? Or I implemented the dynamic version in a wrong way...

@liyaoxuan Did you find the issues with your implementation in your recent commits? If so, is it possible to re-benchmark?

ranshid commented on src/redis-benchmark.c:
```diff
@@ -616,6 +647,44 @@ static void writeHandler(aeEventLoop *el, int fd, void *privdata, int mask) {
         if (config.cluster_mode && c->staglen > 0) setClusterKeyHashTag(c);
         atomicGet(config.slots_last_update, c->slots_last_update);
         c->start = ustime();
+        if (config.qps) {
```
can we encapsulate this qps specific writeHandler logic in a dedicated function (eg. operateQPS or something) right now it makes the writehandler very complicated in appearance

ranshid commented:
@liyaoxuan - would like to ask the status of this change? I believe we all left it aside for some time.

liyaoxuan commented:
@ranshid I was occupied by a project and my course work in the past few months. As the project is almost finished, now I have time to continue working on this feature.

I will try to find the bugs with the version of dynamic threshold in the next few days.

ranshid commented:
@liyaoxuan thank you for this effort!!
I just think we can try and encapsulate the readHandler qps logic and move it to a dedicated function in order to keep better code clarity.

ranshid commented on src/redis-benchmark.c:
```diff
@@ -628,6 +628,50 @@ static void readHandler(aeEventLoop *el, int fd, void *privdata, int mask) {
 
+/* control the QPS of the thread of current client 
+*/
+static int controlQps(void *privdata) {
```
if we are passing it the client anyway, why can't this function just get a pointer to client?

liyaoxuan replied:
you mean the type of parameter of _controlQps()_?

ranshid replied:
> you mean the type of parameter of _controlQps()_?

yes (void *privdata)

liyaoxuan replied:
got it

ranshid reviewed (approved):
LGTM - approved

liyaoxuan commented:
@ranshid I've added some commits to fix bugs, may you review them?

liyaoxuan commented:
@ranshid 
I've added some new commits to fix bugs and hope you can review them, so I click the re-request review button. I don't know if it is a correct operation to ask for review. Apologize if I do it incorrectly.

ranshid commented:
> @ranshid I've added some new commits to fix bugs and hope you can review them, so I click the re-request review button. I don't know if it is a correct operation to ask for review. Apologize if I do it incorrectly.

Thank you @liyaoxuan!
I will make my best to review this ASAP

ranshid commented on src/redis-benchmark.c:
```diff
@@ -1065,21 +1065,25 @@ static void startBenchmarkThreads() {
 static void initQpsControl() {
     int num_threads = config.num_threads == 0 ? 1 : config.num_threads;
-    int qps_per_thread = config.qps / num_threads;
+    int qps_divisor = config.numclients < num_threads ? config.numclients : num_threads;
```
I wonder if we should even allow num_threads > numclients?
What is the logic in that? maybe a better fix would be to change the num_threads to numclients in this case (even when no qps control) and issue a warning on benchmark start?

liyaoxuan replied:
I agree that we should keep num_threads no larger than numclients.
BTW, I wonder what is the correct way to fix this? should I reset and force push or just add a new commit?

ranshid replied:
IMO you can add a new commit

ranshid commented on src/redis-benchmark.c:
```diff
@@ -654,8 +654,9 @@ static int controlQps(client c) {
             if (config.resume_interval[thread_id] - time_elapsed >= 10000 && config.control_granularity[thread_id] >= 2*config.pipeline) {
-                config.resume_interval[thread_id] /= 2;
-                config.control_granularity[thread_id] /= 2;
+                int new_granularity = config.control_granularity[thread_id] / 2;
```
I understand the need for better precision, but maybe we should make use of floating point numbers?
we do it for latency reports so maybe worth investing in this case to?
any way in case we keep it this way maybe we should document it, since it does make the code more complex...

liyaoxuan replied:
I'm not sure if float point number can fix the problem.
The precision problem here affect control effect mainly when the QPS is low. 
For example, assume QPS per threads is 12, the last two group of granularities and resume intervals is (3 250000) and (1 125000) before fixation.
Using float point number makes the last values become (1.5, 125000) and makes no difference as the granularity should be integer.

ranshid replied:
well, I was suggesting that the granularity will also be floated, but I guess it is not a showstopper

liyaoxuan replied:
in line 643  
`if (config.count[thread_id] >= config.control_granularity[thread_id]) {`
1.5 has the same effect as 2, which is also not correct.

ranshid commented on src/redis-benchmark.c:
```diff
@@ -684,6 +684,8 @@ static void writeHandler(aeEventLoop *el, int fd, void *privdata, int mask) {
         int requests_issued = 0;
         atomicGetIncr(config.requests_issued, requests_issued, config.pipeline);
         if (requests_issued >= config.requests) {
+            atomicDecr(config.requests_issued, config.pipeline);
```
Question here: I understand this is a by product of the decrements we do when client is paused right?
So maybe we should have made it differently:
1. in case client is paused we only flag it with pause flag and do not decrement (we are committed to send these requests anyway)
2. on resume we flag the client with resume flag to indicate this is a resumed client.
3. on write handler we will increment only in case the client has no resume flag turned

I admit I did not put extra thought to all cases, but this seems cleaner IMO

liyaoxuan replied:
A little problem here: the number of paused clients may be larger than the number of requests left to be issued.

ranshid replied:
@liyaoxuan, I am not saying the check should not take place. only later when the clients are resumed. that way we can eliminate the decrement code right?

liyaoxuan replied:
The decrement is for those clients that are not killed or paused. When the clients that are responsible for the last few requests are paused, other clients continue calling _writeHandler()_ repeatedly as their file events are not deleted. Thus `config.requests_issued` will be increased over `config.requests` when paused clients are resumed.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/redis/redis?pullRequest=9550) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/redis/redis?pullRequest=9550) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/redis/redis?pullRequest=9550) it.</sub>
